### PR TITLE
fix(zamba):  add use_associative_scan config flag to avoid torch.compile slowdown

### PR DESCRIPTION
**Pull request by msnliu**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48331) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48331)

ZambaMambaMixer.forward() passes use_associative_scan=False directly to mamba_selective_scan, which falls back to a sequential Python loop that makes torch.compile slow.

The other Mamba-family models (mamba, jamba, falcon_mamba) instead read this setting from their config and forward it to the scan function. This PR adds the same plumbing to ZambaConfig and ZambaMambaMixer.

Changes:
1. configuration_zamba.py: add use_associative_scan: bool = True field
2. modeling_zamba.py: read config.use_associative_scan in __init__ and pass it in forward() instead of the hardcoded False

@vasqu

### CODE REVIEW

**Review by vasqu — APPROVED** [review]
Could we have a small fast test to check this in zamba + can you fix the linter (`make style`)

**Comment by msnliu** [comment]
> Could we have a small fast test to check this in zamba + can you fix the linter (`make style`)

Sure, thank you for your prompt review.

**Comment by msnliu** [comment]
> Could we have a small fast test to check this in zamba + can you fix the linter (`make style`)

After looking into it, it seems that `run_common_tests()` already covers this via config save/reload round-trip — it didn't catch it before because the field didn't exist yet. I also checked other Mamba-family models for consistency, and they rely on `run_common_tests()` for this as well. The trailing whitespace has also been fixed from `make style`. Thank you for your time reviewing this.

**Comment by vasqu** [comment]
Ah no for test I meant like running associative vs non associative scan and check that the logits are close enough

**Comment by msnliu** [comment]
> Ah no for test I meant like running associative vs non associative scan and check that the logits are close enough

Got it, I see what you mean now — mamba has `test_associative_scan_matches_sequential` for this. Zamba didn't have one. I'll add the equivalent test and run the comparison. Thanks for the clarification!

**Comment by msnliu** [comment]
> ### CI recap
> **Dashboard:** [View test results in Grafana](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48331) **Latest run:** [33051927474:2](https://github.com/huggingface/transformers/actions/runs/33067498440) **Result:** `failure` | **Jobs:** 5 | **Tests:** 342 | **Failures:** 0 | **Duration:** 42s
> 
> > **Code quality check failed**: test jobs were skipped. Fix the code quality issues and push again to run tests.

Test passed locally (27min15secs, `test_associative_scan_matches_sequential`), make style shows no issues in the changed files. Thanks for reviewing it again.

**Comment by msnliu** [comment]
@vasqu, when you have a chance, could you please review the follow-up for this PR? I'd appreciate your guidance on whether the proposed changes align with the expectations of the Hugging Face community.

File: tests/models/zamba/test_modeling_zamba.py [hunk 1]
```diff
@@ -516,3 +519,36 @@ def test_simple_batched_generate_with_padding(self):
...
+        model = ZambaForCausalLM.from_pretrained(
+            "Zyphra/Zamba-7B-v1", dtype=torch.bfloat16, use_associative_scan=False
+        ).to(torch_device)
+        model.eval()
+        model.forward = torch.compile(model.forward)
+        output = model.generate(input_ids, do_sample=False, use_cache=False, max_new_tokens=10)
+        expected_text = self.tokenizer.decode(output[0].tolist())
+
+        del model
+        torch._dynamo.reset()
```
Comment (vasqu) [inline tests/models/zamba/test_modeling_zamba.py]:
```suggestion
        torch._dynamo.reset()
        torch.cuda.empty_cache()
```

File: tests/models/zamba/test_modeling_zamba.py [hunk 2]
```diff
@@ -516,3 +519,36 @@ def test_simple_batched_generate_with_padding(self):
...
+        # Opt-in: use_associative_scan=True → compiled associative scan
+        model = ZambaForCausalLM.from_pretrained(
+            "Zyphra/Zamba-7B-v1", dtype=torch.bfloat16, use_associative_scan=True
+        ).to(torch_device)
+        model.eval()
+        model.forward = torch.compile(model.forward)
+        output = model.generate(input_ids, do_sample=False, use_cache=False, max_new_tokens=10)
+        output_text = self.tokenizer.decode(output[0].tolist())
+
+        self.assertEqual(output_text, expected_text)
```
Comment (vasqu) [inline tests/models/zamba/test_modeling_zamba.py:554]: I guess this mimics the mamba1 test. SGTM

**Review by vasqu — APPROVED** [review]
Running the slow tests in a sec but would avoid the del model before merge

**Comment by vasqu** [comment]
run-slow: zamba

**Comment by vasqu** [comment]
The AMD failures are irrelevant if you can address my last comments, I think we can merge then :hugs:

File: tests/models/zamba/test_modeling_zamba.py (hunk 2 above)
Comment (msnliu) [reply tests/models/zamba/test_modeling_zamba.py:554]: Yes. I compared the implementation when I drafted this PR.

**Comment by msnliu** [comment]
> The AMD failures are irrelevant if you can address my last comments, I think we can merge then 🤗

Thank you so much for your prompt reply. I have addressed the issue you mentioned. Thanks again for your time and efforts in reviewing this.
